utils/s3/client.cc

```diff
@@ -302,7 +302,7 @@ future<> client::upload_sink::upload_part(unsigned part_number, memory_data_sink
     s3l.trace("PUT part {} {} bytes in {} buffers (upload id {})", part_number, bufs.size(), bufs.buffers().size(), _upload_id);
     auto req = http::request::make("PUT", _client->_host, _object_name);
     req._headers["Content-Length"] = format("{}", bufs.size());
-    req.query_parameters["partNumber"] = format("{}", part_number);
+    req.query_parameters["partNumber"] = format("{}", part_number + 1);
     req.query_parameters["uploadId"] = _upload_id;
```

Comment:
if we think that part_number translates into the actual part number, which starts from 1, then I think it makes sense to adjust client::upload_sink::do_flush() which currently treats part_number like an index, rather than the actual part number.

we just need to do this instead:

```
diff --git a/utils/s3/client.cc b/utils/s3/client.cc
index e3657667e1..268fa377c1 100644
--- a/utils/s3/client.cc
+++ b/utils/s3/client.cc
@@ -329,8 +329,8 @@ future<> client::upload_sink::do_flush() {
     if (!upload_started()) {
         co_await start_upload();
     }
-    auto pn = _part_etags.size();
     _part_etags.emplace_back();
+    auto pn = _part_etags.size();
     co_await upload_part(pn, std::move(_bufs));
 }
 
```

Reply:
In that case `_part_etags[nr]` will off-by-one :)

Reply:
I moved it to #13534

Reply:
I realized later we'd also need adjustment for converting it back into idx, but perhaps part_number should be [1, N] by default, and once we need it as index into _part_etags, then we'd convert it using part_number_to_idx() or something?! My first impression is that part_number refers to the actual upload part number, not an index into an internal structure.